// rsuite/cjs/Drawer/Drawer.js
'use client';
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _Slide = _interopRequireDefault(require("../Animation/Slide"));
var _Modal = _interopRequireDefault(require("../Modal"));
var _hooks = require("../internals/hooks");
var _utils = require("../internals/utils");
var _DrawerContext = _interopRequireDefault(require("./DrawerContext"));
var _propTypes2 = require("../internals/propTypes");
var _DrawerBody = _interopRequireDefault(require("./DrawerBody"));
var _DrawerHeader = _interopRequireDefault(require("./DrawerHeader"));
var _DrawerActions = _interopRequireDefault(require("./DrawerActions"));
var _DrawerFooter = _interopRequireDefault(require("./DrawerFooter"));
var _DrawerTitle = _interopRequireDefault(require("./DrawerTitle"));
var _excluded = ["className", "placement", "classPrefix", "animation", "closeButton"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * The Drawer component is used to display extra content from a main content.
 * @see https://rsuitejs.com/components/drawer
 */
var Drawer = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var className = props.className,
    _props$placement = props.placement,
    placement = _props$placement === void 0 ? 'right' : _props$placement,
    _props$classPrefix = props.classPrefix,
    classPrefix = _props$classPrefix === void 0 ? 'drawer' : _props$classPrefix,
    _props$animation = props.animation,
    animation = _props$animation === void 0 ? _Slide.default : _props$animation,
    _props$closeButton = props.closeButton,
    closeButton = _props$closeButton === void 0 ? true : _props$closeButton,
    rest = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  var _useClassNames = (0, _hooks.useClassNames)(classPrefix),
    merge = _useClassNames.merge,
    prefix = _useClassNames.prefix;
  var classes = merge(className, prefix(placement));
  var animationProps = {
    placement: placement
  };
  var contextValue = (0, _react.useMemo)(function () {
    return {
      closeButton: closeButton,
      isDrawer: true
    };
  }, [closeButton]);
  return /*#__PURE__*/_react.default.createElement(_DrawerContext.default.Provider, {
    value: contextValue
  }, /*#__PURE__*/_react.default.createElement(_Modal.default, (0, _extends2.default)({}, rest, {
    ref: ref,
    overflow: false,
    classPrefix: classPrefix,
    className: classes,
    animation: animation,
    animationProps: animationProps
  })));
});
_DrawerBody.default.displayName = 'DrawerBody';
_DrawerHeader.default.displayName = 'DrawerHeader';
_DrawerActions.default.displayName = 'DrawerActions';
_DrawerFooter.default.displayName = 'DrawerFooter';
_DrawerTitle.default.displayName = 'DrawerTitle';
Drawer.Body = _DrawerBody.default;
Drawer.Header = _DrawerHeader.default;
Drawer.Actions = _DrawerActions.default;
Drawer.Footer = (0, _utils.deprecateComponent)(_DrawerFooter.default, '<Drawer.Footer> has been deprecated, use <Drawer.Actions> instead.');
Drawer.Title = _DrawerTitle.default;
Drawer.displayName = 'Drawer';
Drawer.propTypes = (0, _extends2.default)({}, _Modal.default.propTypes, {
  classPrefix: _propTypes.default.string,
  placement: (0, _propTypes2.oneOf)(['top', 'right', 'bottom', 'left']),
  children: _propTypes.default.node,
  className: _propTypes.default.string
});
var _default = exports.default = Drawer;
