// rsuite/cjs/Text/Text.js
'use client';
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _hooks = require("../internals/hooks");
var _propTypes2 = require("../internals/propTypes");
var _excluded = ["as", "align", "classPrefix", "color", "className", "maxLines", "weight", "muted", "transform", "size", "style"];
var fontSizeMap = {
  sm: 12,
  md: 14,
  lg: 16,
  xl: 18,
  xxl: 20
};
/**
 *
 * The `Text` component is used to display text.
 *
 * @see https://rsuitejs.com/components/text
 */
var Text = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _props$as = props.as,
    Component = _props$as === void 0 ? 'p' : _props$as,
    align = props.align,
    _props$classPrefix = props.classPrefix,
    classPrefix = _props$classPrefix === void 0 ? 'text' : _props$classPrefix,
    color = props.color,
    className = props.className,
    maxLines = props.maxLines,
    weight = props.weight,
    muted = props.muted,
    transform = props.transform,
    size = props.size,
    style = props.style,
    rest = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  var _useClassNames = (0, _hooks.useClassNames)(classPrefix),
    withClassPrefix = _useClassNames.withClassPrefix,
    merge = _useClassNames.merge;
  var classes = merge(className, withClassPrefix(color, align, weight, transform, {
    muted: muted,
    ellipsis: maxLines
  }));
  var styles = (0, _extends2.default)({
    fontSize: fontSizeMap[size] || size
  }, maxLines ? {
    WebkitLineClamp: maxLines
  } : null, style);
  return /*#__PURE__*/_react.default.createElement(Component, (0, _extends2.default)({}, rest, {
    ref: ref,
    className: classes,
    style: styles
  }));
});
Text.displayName = 'Text';
Text.propTypes = {
  className: _propTypes.default.string,
  classPrefix: _propTypes.default.string,
  as: _propTypes.default.elementType,
  size: _propTypes.default.oneOfType([_propTypes.default.number, (0, _propTypes2.oneOf)(['sm', 'md', 'lg', 'xl', 'xxl'])]),
  muted: _propTypes.default.bool,
  transform: (0, _propTypes2.oneOf)(['uppercase', 'lowercase', 'capitalize']),
  align: (0, _propTypes2.oneOf)(['left', 'center', 'right', 'justify']),
  weight: (0, _propTypes2.oneOf)(['thin', 'light', 'regular', 'medium', 'semibold', 'bold', 'extrabold']),
  maxLines: _propTypes.default.number
};
var _default = exports.default = Text;
